File: crates/agent-loop/src/events.rs
//! Event plumbing for the agent loop: the turn-level LoopEvent and the
//! optional EventSink injected into a DefaultAgentLoop.
//!
//! Kept separate from the loop driver so the event type and its sink can be
//! used (and unit-tested) without pulling in the whole turn loop.

use std::sync::Arc;

use serde_json::Value;
use celestea_core::{Message, ToolOutput, TurnOutcome};

/// An event a running turn delivers to an injected sink. This is the
/// agent-loop view of the turn: it carries the LLM stream events
/// (Text/Thinking/Done, mirroring StreamEvent) plus the tool lifecycle
/// (ToolCall + the full ToolOutput for its ToolResult), so a rich UI can draw
/// tool cards and style thinking without scraping the session log or relying
/// on StreamEvent growing variants (core is frozen for P1).
///
/// P0-A: every started turn emits exactly ONE terminal [LoopEvent::TurnEnd]
/// carrying the real terminal state — consumers map it onto their "done"
/// envelope (SSE done with an outcome field), so cancelled / error /
/// step-limit / interrupted turns can never be mistaken for completed ones.
#[derive(Debug, Clone)]
pub enum LoopEvent {
    Text(String),
    Thinking(String),
    /// The authoritative assistant reply of one model step (not terminal:
    /// a tool-call step is followed by more events).
    Done(Message),
    ToolCall { id: String, name: String, args: Value },
    ToolResult(ToolOutput),
    /// Terminal turn verdict, emitted exactly once per turn on every path
    /// (completed / cancelled / error / step_limit / interrupted).
    TurnEnd(TurnOutcome),
}

/// A sink receives every LoopEvent a turn produces, in log order. None on
/// the loop (the default) keeps the legacy behavior: Text/Thinking deltas are
/// printed straight to stdout.
pub type EventSink = Arc<dyn Fn(LoopEvent) + Send + Sync>;

File: packages/agent-loop/src/events.ts
/**
 * Turn-level event plumbing — port of `crates/agent-loop/src/events.rs`.
 *
 * A running turn delivers every [LoopEvent] it produces to an injected
 * [EventSink], in log order: the LLM stream deltas (text / thinking / done)
 * plus the tool lifecycle (tool_call + the FULL ToolOutput of every result) so
 * a rich UI can draw tool cards without scraping the session log.
 *
 * P0-A: every started turn emits exactly ONE terminal
 * `{ kind: "turn_end" }` carrying the real terminal state — consumers map it
 * onto their "done"/status envelope, so cancelled / error / step-limit /
 * interrupted turns can never be mistaken for completed ones.
 *
 * The builders below are the only place where core values become events, so
 * the log and the event stream can never drift apart.
 */

import {
  messageTexts,
  messageToolCalls,
  type LoopEvent,
  type Message,
  type ToolCall,
  type ToolDecision,
  type ToolOutput,
  type TurnOutcome,
} from "@celestea/core";

/** A sink receives every LoopEvent of a turn, in log order. */
export type EventSink = (event: LoopEvent) => void;

/** `Some(ToolDecision::Allow)` -> `"allow"`: the flat label of the SSE payload. */
export function decisionLabel(decision: ToolDecision | null): "allow" | "deny" | "ask" | null {
  return decision === null ? null : decision.kind;
}

export function toolCallEvent(call: ToolCall): LoopEvent {
  return { kind: "tool_call", id: call.id, name: call.name, args: call.args };
}

/** The full ToolOutput rides the event: value, authored render, error, verdict. */
export function toolResultEvent(output: ToolOutput): LoopEvent {
  return {
    kind: "tool_result",
    callId: output.call_id,
    ok: output.error === null,
    value: output.value,
    render: output.render,
    error: output.error,
    decision: decisionLabel(output.decision),
  };
}

/** The authoritative assistant reply of one model step (not terminal). */
export function doneEvent(message: Message): LoopEvent {
  return {
    kind: "done",
    text: messageTexts(message).join(""),
    tool_calls: messageToolCalls(message).map((call) => ({ id: call.id, name: call.name, args: call.args })),
  };
}

/** The single terminal verdict of the turn. */
export function turnEndEvent(outcome: TurnOutcome): LoopEvent {
  return { kind: "turn_end", outcome };
}
